src/EnergyPlus/PlantLoopHeatPumpEIR.cc

```diff
@@ -1289,4 +1290,1373 @@ void EIRPlantLoopHeatPump::oneTimeInit(EnergyPlusData &state)
+    return nullptr; // LCOV_EXCL_LINE
+}
+
+void EIRFuelFiredHeatPump::pairUpCompanionCoils(EnergyPlusData &state)
+{
+    for (auto &thisHP : state.dataEIRFuelFiredHeatPump->heatPumps) {
+        if (!thisHP.companionCoilName.empty()) {
+            auto thisCoilName = UtilityRoutines::MakeUPPERCase(thisHP.name);
+            auto &thisCoilType = thisHP.EIRHPType;
+            auto targetCompanionName = UtilityRoutines::MakeUPPERCase(thisHP.companionCoilName);
+            for (auto &potentialCompanionCoil : state.dataEIRFuelFiredHeatPump->heatPumps) {
+                auto &potentialCompanionType = potentialCompanionCoil.EIRHPType;
+                auto potentialCompanionName = UtilityRoutines::MakeUPPERCase(potentialCompanionCoil.name);
+                if (potentialCompanionName == thisCoilName) {
+                    // skip the current coil
+                    continue;
+                }
+                if (potentialCompanionName == targetCompanionName) {
+                    if (thisCoilType == potentialCompanionType) {
+                        ShowSevereError(state,
+                                        "Invalid companion specification for EIR Plant Loop Fuel-Fired Heat Pump named \"" + thisCoilName + "\"");
+                        ShowContinueError(state, "For heating objects, the companion must be a cooling object, and vice-versa");
+                        ShowFatalError(state, "Invalid companion object causes program termination");
+                    }
+                    thisHP.companionHeatPumpCoil = &potentialCompanionCoil;
+                    break;
+                }
+            }
+            if (!thisHP.companionHeatPumpCoil) {
+                ShowSevereError(state, "Could not find matching companion heat pump coil.");
+                ShowContinueError(state, "Base coil: " + thisCoilName);
+                ShowContinueError(state, "Looking for companion coil named: " + targetCompanionName);
+                ShowFatalError(state, "Simulation aborts due to previous severe error");
+            }
+        }
+    }
+}
+
+void EIRFuelFiredHeatPump::processInputForEIRPLHP(EnergyPlusData &state)
+{
+    struct ClassType
+    {
+        DataPlant::PlantEquipmentType thisType;
+        std::string nodesType;
+        std::function<Real64(Real64, Real64)> calcLoadOutletTemp;
+        std::function<Real64(Real64, Real64)> calcQsource;
+        std::function<Real64(Real64, Real64)> calcSourceOutletTemp;
+
+        ClassType(DataPlant::PlantEquipmentType _thisType,
+                  std::string _nodesType,
+                  std::function<Real64(Real64, Real64)> _tLoadOutFunc,
+                  std::function<Real64(Real64, Real64)> _qSrcFunc,
+                  std::function<Real64(Real64, Real64)> _tSrcOutFunc)
+            : thisType(_thisType), nodesType(std::move(_nodesType)), calcLoadOutletTemp(std::move(_tLoadOutFunc)), calcQsource(std::move(_qSrcFunc)),
```

Comment:
There is no need to `std::move` function pointers. There is no such thing as deep copying a function pointer, and by default pointers are not deep copied like references are.

Reply:
ping @jcyuan2020

Reply:
I think I roughly get the idea. So the changed code would be like: 

```
        ClassType(DataPlant::PlantEquipmentType _thisType,
                  std::string _nodesType,
                  std::function<Real64(Real64, Real64)> _tLoadOutFunc,
                  std::function<Real64(Real64, Real64)> _qSrcFunc,
                  std::function<Real64(Real64, Real64)> _tSrcOutFunc)
            : thisType(_thisType), nodesType(std::move(_nodesType)), calcLoadOutletTemp(_tLoadOutFunc), calcQsource(_qSrcFunc), calcSourceOutletTemp(_tSrcOutFunc)
        {
        }
```

And the `nodesType(std::move(_nodesType))` one should be kept as it is, since this might involve a "deep copy" for the std::string type? (if say, that in the previous comments, that this `std::string nodesType` is still used rather than designing an enum type).